sklearn/linear_model/coordinate_descent.py

```diff
@@ -2159,3 +2161,315 @@ def __init__(self, eps=1e-3, n_alphas=100, alphas=None, fit_intercept=True,
+                    return (- x ** 2 / self.alpha +
+                            2 * self.q * x - self.alpha) / \
+                            (2 * (self.q - 1))
+                else:
+                    return (self.q + 1) * self.alpha / 2
+            p = np.vectorize(p, ['float'])
+        return p(x)
+
+    def _p_prime(self, x):
+        """Calculate penalty term derivative."""
+        if self.penalty == 'lq':
+            def p_prime(x):
+                return self.q / (np.power(x, 1 - self.q) + self.eps)
+        elif self.penalty == 'log':
+            def p_prime(x):
+                return 1. / (x + self.eps)
+        elif self.penalty == 'scad':
+            def p_prime(x):
+                if x <= self.alpha:
+                    return 1.
+                elif x < self.q * self.alpha:
+                    return (- x / self.alpha + self.q) / \
+                            (self.q - 1)
+                else:
+                    return self.eps
+            p_prime = np.vectorize(p_prime, ['float'])
+        return p_prime(x)
+
+    def loss_(self, pred, y):
+        """Calculate model loss."""
+        return 1. / (2 * y.shape[0]) \
+            * np.sum(np.square(y - pred)) \
+            + self.alpha * np.sum(self._p(np.abs(self.coef_)))
+
+    def fit(self, X, y, check_input=True):
+        """
+        Fit Lasso models, each with coordinate descent.
+
+        Parameters
+        -----------
+        X : ndarray or scipy.sparse matrix, shape (n_samples, n_features)
+            Data
+
+        y : ndarray, shape (n_samples,)
+            Target
+        """
+        # We expect X and y to be already float64 Fortran ordered arrays
+        # when bypassing checks
+        if check_input:
+            y = np.asarray(y, dtype=np.float64)
+            X, y = check_X_y(X, y, accept_sparse='csc', dtype=np.float64,
+                             order='F',
+                             copy=self.copy_X and self.fit_intercept,
+                             multi_output=True, y_numeric=True)
+            y = check_array(y, dtype=np.float64, order='F', copy=False,
+                            ensure_2d=False)
+        X, y, X_offset, y_offset, X_scale, precompute, Xy = \
+            _pre_fit(X, y, None, self.precompute, self.normalize,
+                     self.fit_intercept, copy=False)
+
+        y = column_or_1d(y, warn=True)
+
+        n_samples, n_features = X.shape
+        X_sparse = sparse.isspmatrix(X)
+
+        self._check_params()
+        self.coef_ = np.zeros(n_features, dtype=np.float64, order='F')
+        self.train_score_ = np.zeros(self.max_lasso_iterations)
+        weights = np.ones(n_features)
+
+        # Initial Lasso fit
+        super(AdaptiveLasso, self).fit(X, y)
+        self.train_score_[0] = self.loss_(y, self.predict(X))
+
+        for k in xrange(1, self.max_lasso_iterations):
+            self.n_iter_ = k
+            weights = self._p_prime(np.abs(self.coef_))
```

Comment:
This weighted fitting feels like it could potentially live in the parent class, since weights are not specific to the Adaptive Lasso but occur in other variant of the Lasso; the Adaptive Lasso is unique for how and when it chooses weights, not that it has weights.